### PR TITLE
[addons] fix usage of CVariant map after 8851244

### PR DESCRIPTION
**Pull request by Rechi**
## Description
Why this is needed: https://github.com/xbmc/xbmc/pull/12244#pullrequestreview-42553186

## Motivation and Context
Fix crash caused by 8851244 (#12244)
@ksooo please check if this solves your crash

## How Has This Been Tested?
runtime tested, now crash is gone

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by tamland** [comment]
I don't think this is correct.. At least this is not how std containers work (you can iterate an empty map just fine)

**Comment by Rechi** [comment]
You are partly correct, normally the empty check for pure std::map isn't necessary, but here we interact with `CVariant`.
Have a look at https://github.com/xbmc/xbmc/blob/d8fbb5a/xbmc/utils/Variant.cpp#L784-L814, especially the else branches. If art wasn't parsed into the CVariant you get different iterators for begin_map and end_map and you will iterate over unexpected memory locations.
The real issue is in the CVariant class. I'll have a look if it can be fixed there easily, which is of course the preferred way.

File: xbmc/addons/AddonDatabase.cpp [hunk 1]
```diff
@@ -99,8 +99,11 @@ static void DeserializeMetadata(const std::string& document, CAddonBuilder& buil
   builder.SetIcon(variant["icon"].asString());
 
   std::map<std::string, std::string> art;
-  for (auto it = variant["art"].begin_map(); it != variant["art"].end_map(); ++it)
-    art.emplace(it->first, it->second.asString());
+  if (!variant["art"].empty())
```
Comment (Montellese) [inline xbmc/addons/AddonDatabase.cpp]: The proper check is `if (variant.isMember("art") && variant["art"].isObject())`. Then you  can just iterate over it even if the "art" member is an empty map.

**Comment by Montellese** [comment]
The correct usage of CVariant is to first check if a member exists using `isMember()` and if the type of the member is what you expect it to be. You can't just assume that it's an object/map and start iterating over it with `begin_map()`. CVariant wasn't designed to just magically work like a python dict.

**Comment by tamland** [comment]
> You can't just assume that it's an object/map and start iterating over it with begin_map(). CVariant wasn't designed to just magically work like a python dict.

It definitely *reads* like it should as every other method in CVariant returns the empty variant when the type is 'wrong'. So I expected the empty variant to also return an empty iterator, not undefined behavior. I mean what's event the point of returning an iterator for unexpected types if it cannot be used?

**Comment by Rechi** [comment]
`begin_map()` and `end_map()` both return an empty iterator, but not the same.
Problem now is, begin != end, and therefore it iterates over unexpected memory locations.

#12264 would correct the behaviour to return the same iterator for begin and end, but I'm not sure what @Montellese thinks about that.

**Comment by Rechi** [comment]
fixed with #12264

**Pull request closed without merging**